src/main/java/com/google/devtools/build/lib/remote/GrpcCacheClient.java

```diff
@@ -345,12 +352,41 @@ public static String getResourceName(String instanceName, Digest digest) {
                 .setReadOffset(offset.get())
                 .build(),
             new StreamObserver<ReadResponse>() {
+              InputStream inner;
+              ZstdInputStream zis;
+
+              {
+                initialise();
+              }
+
+              private void initialise() throws IOException {
+                if (options.cacheByteStreamCompression) {
+                  zis =
+                      new ZstdInputStream(
+                          new InputStream() {
+                            @Override
+                            public int read() throws IOException {
+                              return inner.read();
+                            }
+                          });
+
+                  zis.setContinuous(true);
+                }
+              }
+
               @Override
               public void onNext(ReadResponse readResponse) {
                 ByteString data = readResponse.getData();
                 try {
-                  data.writeTo(out);
-                  offset.addAndGet(data.size());
+                  if (options.cacheByteStreamCompression) {
```

Comment:
Instead of handling two cases here, can we wrap the output stream ahead of time if compression is enabled e.g.
```
if (options.cacheByteStreamCompression) {
    out = decompress(out);
}
```

so whenever we write to the new `out`, the content is decompressed and redirected to the original `out`.

With that, we can keep the code here unchanged.